server/services/RunService.js

```diff
@@ -38,13 +38,281 @@ const db = require('../models/index');
+            }
+        }
+        const updateActiveRuns = [];
+        const updateInactiveRuns = [];
+        for (let existingRun of existingRuns) {
+            let matchIndex = addRuns.findIndex(r => {
+                return (
+                    r.apg_example_id === existingRun.ApgExample.id
+                    && r.browser_version_to_at_version === existingRun.BrowserVersionToAtVersion.id
+                );
+            });
+
+            // Remove the element because you don't need ot add just add/update
+            if (matchIndex) {
+                addRuns.splice(matchIndex, 1);
+
+                if (!existingRun.active) {
+                    updateActiveRuns.push(existingRun);
+                }
+            }
+            else if (existingRun.active) {
+                updateInactiveRuns.push(existingRun);
+            }
+        }
+
+        if (updateActiveRuns.length) {
+            let ids = updateActiveRuns.map(t => t.id);
+            await db.Run.update(
+                { active: true },
+                { where: { id: ids } }
+            );
+        }
+
+        if (updateInactiveRuns.length) {
+            let ids = updateInactiveRuns.map(t => t.id);
+            await db.Run.update(
+                { active: false },
+                { where: { id: ids } }
+            );
+        }
+
+
+        let runStatus = await db.RunStatus.findOne({
+            where: { name: 'raw' }
+        });
+
+        // TODO, not sure if we should remove this
+        let ats = await db.At.findAll({
+            where: { test_version_id }
+        });
+        let atNameToAt = ats.reduce((acc, curr) => {
+            acc[curr.at_name_id] = curr.id;
+            return acc;
+        }, {});
+
+        // TODO, remove this:
+        let user = await db.Users.findOne();
+        if (!user) {
+            let user = await db.Users.create({});
+        }
+        const testCycles = await db.TestCycle.findCreateFind({
+            where: {
+                test_version_id,
+                created_user_id: user.id
+            }
+        });
+
+        if (addRuns.length) {
+            let dbRuns = addRuns.map(r => {
+                let at_id = atNameToAt[r.at_name_id];
+                return {
+                    browser_version_id: r.browser_version_id, // eventually will remove column
+                    at_version_id: r.at_version_id,           // eventually will remove column
+                    test_cycle_id: testCycles[0].id,              // eventually will remove column
+                    at_id,                                    // maybe eventually will remove column
+                    browser_version_to_at_versions_id: r.browser_version_to_at_version_id,
+                    apg_example_id: r.apg_example_id,
+                    test_version_id,
+                    active: true,
+                    run_status_id: runStatus.id
+                };
+            });
+
+            await db.Run.bulkCreate(dbRuns);
```

Comment:
This is the only time you do anything in bulk. But you have a lot of loops above where I think you could/should be doing bulk actions.